Automated Smoketest: :x:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: d584a5b4066a56a22088c6f6e554bd70941326ca
Number of Error(s): 29
/mnt/disk1/home/vamosax/smoketest/PR-12000/HPCC-Platform/ecl/hthor/hthor.cpp:8270:1: error: expected primary-expression before \xe2\x80\x98<<\xe2\x80\x99 token

[Build error:](https://gist.github.com/78c946994d9f4446c203a0b67b85c5e4)
Build log [Build log](https://gist.github.com/HPCCSmoketest/78c946994d9f4446c203a0b67b85c5e4#file-result-18-12-14-07-54-44-log)

--- a/ecl/hthor/hthor.cpp
+++ b/ecl/hthor/hthor.cpp
@@ -8266,9 +8266,43 @@ bool CHThorDiskReadBaseActivity::openNext()
                 actualFilter.appendFilters(fieldFilters);
 
             bool canSerializeTypeInfo = actualDiskMeta->queryTypeInfo()->canSerialize() && projectedDiskMeta->queryTypeInfo()->canSerialize();
+
+            /* If part can potentially be remotely streamed, 1st check if any part is local,
+             * then try to remote stream, and otherwise failover to legacy remote access
+             */
+            unsigned localCopy = NotFound;
             if (canSerializeTypeInfo && (rt_binary == readType))
             {
-                for (unsigned copy=0; copy < numCopies; copy++)
+                std::vector<unsigned> remoteCandidates;
+                // scan for local part 1st
+                for (unsigned copy=0; copy<numCopies; copy++)
+                {
+                    RemoteFilename rfn;
+                    curPart->getFilename(rfn, copy);
+                    StringBuffer localPath;
+                    if (!isRemoteReadCandidate(agent, rfn, localPath))
+                    {
+                        Owned<IFile> iFile = createIFile(localPath.str());
+                        try
+                        {
+                            if (iFile->exists())
+                            {
+                                localCopy = copy;
+                                remoteCandidates.clear();
+                                break;
+                            }
+                        }
+                        catch (IException *e)
+                        {
+                            EXCLOG(e, "CHThorDiskReadBaseActivity::openNext()");
+                            e->Release();
+                        }
+                    }
+                    else
+                        remoteCandidates.push_back(copy);
+                }
+
+                for (unsigned &copy: remoteCandidates)
                 {
                     RemoteFilename rfilename;
                     if (curPart)
@@ -8281,42 +8315,43 @@ bool CHThorDiskReadBaseActivity::openNext()
                     {
                         // NB: only binary handles can be remotely processed by dafilesrv at the moment
 
-                        StringBuffer path;
-                        if (isRemoteReadCandidate(agent, rfilename, path))
+                        // Open a stream from remote file, having passed actual, expected, projected, and filters to it
+                        SocketEndpoint ep(rfilename.queryEndpoint());
+                        setDafsEndpointPort(ep);
+
+                        StringBuffer localPath;
+                        rfilename.getLocalPath(localPath);
+                        Owned<IRemoteFileIO> remoteFileIO = createRemoteFilteredFile(ep, localPath, actualDiskMeta, projectedDiskMeta, actualFilter, compressed, grouped, remoteLimit);
+                        if (remoteFileIO)
                         {
-                            // Open a stream from remote file, having passed actual, expected, projected, and filters to it
-                            SocketEndpoint ep(rfilename.queryEndpoint());
-                            setDafsEndpointPort(ep);
+                            StringBuffer tmp;
+                            remoteFileIO->addVirtualFieldMapping("logicalFilename", logicalFileName.str());
+                            remoteFileIO->addVirtualFieldMapping("baseFpos", tmp.clear().append(offsetOfPart).str());
+                            remoteFileIO->addVirtualFieldMapping("partNum", tmp.clear().append(curPart->getPartIndex()).str());
 
-                            Owned<IRemoteFileIO> remoteFileIO = createRemoteFilteredFile(ep, path, actualDiskMeta, projectedDiskMeta, actualFilter, compressed, grouped, remoteLimit);
-                            if (remoteFileIO)
+                            try
                             {
-                                StringBuffer tmp;
-                                remoteFileIO->addVirtualFieldMapping("logicalFilename", logicalFileName.str());
-                                remoteFileIO->addVirtualFieldMapping("baseFpos", tmp.clear().append(offsetOfPart).str());
-                                remoteFileIO->addVirtualFieldMapping("partNum", tmp.clear().append(curPart->getPartIndex()).str());
+                                remoteFileIO->ensureAvailable(); // force open now, because want to failover to other copies or legacy if fails
+                            }
+                            catch (IException *e)
+                            {
+                                EXCLOG(e, nullptr);
+                                e->Release();
+                                continue; // try next copy and ultimately failover to local when no more copies
+                            }
 
-                                try
-                                {
-                                    remoteFileIO->ensureAvailable(); // force open now, because want to failover to other copies or legacy if fails
-                                }
-                                catch (IException *e)
-                                {
-                                    EXCLOG(e, nullptr);
-                                    e->Release();
-                                    continue; // try next copy and ultimately failover to local when no more copies
-                                }
+                            Owned<IFile> iFile = createIFile(rfilename);
 
-                                inputfile.setown(createIFile(rfilename));
-
-                                actualDiskMeta.set(projectedDiskMeta);
-                                expectedDiskMeta = projectedDiskMeta;
-                                actualFilter.clear();
-                                inputfileio.setown(remoteFileIO.getClear());
+                            actualDiskMeta.set(projectedDiskMeta);
+                            expectedDiskMeta = projectedDiskMeta;
+                            actualFilter.clear();
+                            inputfileio.setown(remoteFileIO.getClear());
+                            if (inputfileio)
+                            {
+                                inputfile.setown(iFile.getClear());
+                                break;
                             }
                         }
-                        if (inputfileio)
-                            break;
                     }
                     catch (IException *E)
                     {
@@ -8330,7 +8365,9 @@ bool CHThorDiskReadBaseActivity::openNext()
             }
             if (!inputfile)
             {
-                for (unsigned copy=0; copy < numCopies; copy++)
+                unsigned startCopy = (NotFound != localCopy) ? localCopy : 0;
+                unsigned copy = startCopy;
+                while (true)
                 {
                     RemoteFilename rfilename;
                     if (curPart)
@@ -8368,6 +8405,10 @@ bool CHThorDiskReadBaseActivity::openNext()
                         else
                             saveOpenExc.setown(E);
                     }
+                    if (++copy == numCopies) // wrap
+                        copy = 0;
+                    if (copy == startCopy) // reached starting copy, so scanned all and failed to open any.
+                        break;
                 }
             }
 

--- a/thorlcr/activities/diskread/thdiskreadslave.cpp
+++ b/thorlcr/activities/diskread/thdiskreadslave.cpp
@@ -256,57 +256,87 @@ void CDiskRecordPartHandler::open()
     Owned<ITranslator> translator = activity.getTranslators(*partDesc);
     IOutputMetaData *actualFormat = translator ? &translator->queryActualFormat() : expectedFormat;
     bool canSerializeTypeInfo = actualFormat->queryTypeInfo()->canSerialize() && projectedFormat->queryTypeInfo()->canSerialize();
+
+    /* If part can potentially be remotely streamed, 1st check if any part is local,
+     * then try to remote stream, and otherwise failover to legacy remote access
+     */
     if (canSerializeTypeInfo)
     {
+        std::vector<unsigned> remoteCandidates;
+        // scan for non remote candidate part 1st
         for (unsigned copy=0; copy<partDesc->numCopies(); copy++)
+        {
+            RemoteFilename rfn;
+            partDesc->getFilename(copy, rfn);
+            StringBuffer localPath;
+            if (!isRemoteReadCandidate(activity, rfn, localPath))
+            {
+                Owned<IFile> iFile = createIFile(localPath.str());
+                try
+                {
+                    if (iFile->exists())
+                    {
+                        remoteCandidates.clear();
+                        break;
+                    }
+                }
+                catch (IException *e)
+                {
+                    ActPrintLog(&activity, e, "CDiskRecordPartHandler::open()");
+                    e->Release();
+                }
+            }
+            else
+                remoteCandidates.push_back(copy);
+        }
+        for (unsigned &copy: remoteCandidates) // only if no local part found above
         {
             RemoteFilename rfn;
             partDesc->getFilename(copy, rfn);
 
             StringBuffer path;
-            if (isRemoteReadCandidate(activity, rfn, path))
+            rfn.getLocalPath(path);
+            // Open a stream from remote file, having passed actual, expected, projected, and filters to it
+            SocketEndpoint ep(rfn.queryEndpoint());
+            setDafsEndpointPort(ep);
+
+            RowFilter actualFilter;
+            if (activity.fieldFilters.ordinality())
             {
-                // Open a stream from remote file, having passed actual, expected, projected, and filters to it
-                SocketEndpoint ep(rfn.queryEndpoint());
-                setDafsEndpointPort(ep);
+                if (actualFormat != expectedFormat && translator->queryKeyedTranslator())
+                    translator->queryKeyedTranslator()->translate(actualFilter, activity.fieldFilters);
+                else
+                    actualFilter.appendFilters(activity.fieldFilters);
+            }
+            Owned<IRemoteFileIO> iRemoteFileIO = createRemoteFilteredFile(ep, path, actualFormat, projectedFormat, actualFilter, compressed, activity.grouped, activity.remoteLimit);
+            if (iRemoteFileIO)
+            {
+                StringBuffer tmp;
+                iRemoteFileIO->addVirtualFieldMapping("logicalFilename", logicalFilename.get());
+                iRemoteFileIO->addVirtualFieldMapping("baseFpos", tmp.clear().append(fileBaseOffset).str());
+                iRemoteFileIO->addVirtualFieldMapping("partNum", tmp.clear().append(partDesc->queryPartIndex()).str());
+                rfn.getPath(path);
+                filename.set(path);
+                checkFileCrc = false;
 
-                RowFilter actualFilter;
-                if (activity.fieldFilters.ordinality())
+                try
                 {
-                    if (actualFormat != expectedFormat && translator->queryKeyedTranslator())
-                        translator->queryKeyedTranslator()->translate(actualFilter, activity.fieldFilters);
-                    else
-                        actualFilter.appendFilters(activity.fieldFilters);
+                    iRemoteFileIO->ensureAvailable(); // force open now, because want to failover to other copies if fails (e.g. remote part is missing)
                 }
-                Owned<IRemoteFileIO> iRemoteFileIO = createRemoteFilteredFile(ep, path, actualFormat, projectedFormat, actualFilter, compressed, activity.grouped, activity.remoteLimit);
-                if (iRemoteFileIO)
+                catch (IException *e)
                 {
-                    StringBuffer tmp;
-                    iRemoteFileIO->addVirtualFieldMapping("logicalFilename", logicalFilename.get());
-                    iRemoteFileIO->addVirtualFieldMapping("baseFpos", tmp.clear().append(fileBaseOffset).str());
-                    iRemoteFileIO->addVirtualFieldMapping("partNum", tmp.clear().append(partDesc->queryPartIndex()).str());
-                    rfn.getPath(path);
-                    filename.set(path);
-                    checkFileCrc = false;
-
-                    try
-                    {
-                        iRemoteFileIO->ensureAvailable(); // force open now, because want to failover to other copies if fails (e.g. remote part is missing)
-                    }
-                    catch (IException *e)
-                    {
-                        EXCLOG(e, nullptr);
-                        e->Release();
-                        continue; // try next copy and ultimately failover to local when no more copies
-                    }
-                    partStream.setown(createRowStreamEx(iRemoteFileIO, activity.queryProjectedDiskRowInterfaces(), 0, (offset_t)-1, (unsigned __int64)-1, rwFlags, nullptr, this));
-                    ActPrintLog(&activity, "%s[part=%d]: reading remote dafilesrv file '%s' (logical file = %s)", kindStr, which, path.str(), activity.logicalFilename.get());
-                    break;
+                    EXCLOG(e, nullptr);
+                    e->Release();
+                    continue; // try next copy and ultimately failover to local when no more copies
                 }
+                partStream.setown(createRowStreamEx(iRemoteFileIO, activity.queryProjectedDiskRowInterfaces(), 0, (offset_t)-1, (unsigned __int64)-1, rwFlags, nullptr, this));
+                ActPrintLog(&activity, "%s[part=%d]: reading remote dafilesrv file '%s' (logical file = %s)", kindStr, which, path.str(), activity.logicalFilename.get());
+                break;
             }
         }
     }
 
+    // either local file was found, or no streamable remote parts
     if (!partStream)
     {
         CDiskPartHandlerBase::open(); // NB: base opens an IFile

--- a/thorlcr/activities/indexread/thindexreadslave.cpp
+++ b/thorlcr/activities/indexread/thindexreadslave.cpp
@@ -163,61 +163,89 @@ public:
 
                 unsigned crc=0;
                 part.getCrc(crc);
+
+                /* If part can potentially be remotely streamed, 1st check if any part is local,
+                 * then try to remote stream, and otherwise failover to legacy remote access
+                 */
                 if (canSerializeTypeInfo && !usesBlobs && !localMerge)
                 {
+                    std::vector<unsigned> remoteCandidates;
                     for (unsigned copy=0; copy<part.numCopies(); copy++)
+                    {
+                        RemoteFilename rfn;
+                        part.getFilename(copy, rfn);
+                        StringBuffer localPath;
+                        if (!isRemoteReadCandidate(*this, rfn, localPath))
+                        {
+                            Owned<IFile> iFile = createIFile(localPath.str());
+                            try
+                            {
+                                if (iFile->exists())
+                                {
+                                    remoteCandidates.clear();
+                                    break;
+                                }
+                            }
+                            catch (IException *e)
+                            {
+                                ActPrintLog(e, "getNextInput()");
+                                e->Release();
+                            }
+                        }
+                        else
+                            remoteCandidates.push_back(copy);
+                    }
+                    for (unsigned &copy : remoteCandidates) // only if no local part found above
                     {
                         RemoteFilename rfn;
                         part.getFilename(copy, rfn);
                         StringBuffer path;
                         rfn.getPath(path);
 
-                        StringBuffer lPath;
-                        if (isRemoteReadCandidate(*this, rfn, lPath))
+                        // Open a stream from remote file, having passed actual, expected, projected, and filters to it
+                        SocketEndpoint ep(rfn.queryEndpoint());
+                        setDafsEndpointPort(ep);
+
+                        IConstArrayOf<IFieldFilter> fieldFilters;  // These refer to the expected layout
+                        struct CIndexReadContext : implements IIndexReadContext
                         {
-                            // Open a stream from remote file, having passed actual, expected, projected, and filters to it
-                            SocketEndpoint ep(rfn.queryEndpoint());
-                            setDafsEndpointPort(ep);
-
-                            IConstArrayOf<IFieldFilter> fieldFilters;  // These refer to the expected layout
-                            struct CIndexReadContext : implements IIndexReadContext
+                            IConstArrayOf<IFieldFilter> &fieldFilters;
+                            CIndexReadContext(IConstArrayOf<IFieldFilter> &_fieldFilters) : fieldFilters(_fieldFilters)
                             {
-                                IConstArrayOf<IFieldFilter> &fieldFilters;
-                                CIndexReadContext(IConstArrayOf<IFieldFilter> &_fieldFilters) : fieldFilters(_fieldFilters)
-                                {
-                                }
-                                virtual void append(IKeySegmentMonitor *segment) override { throwUnexpected(); }
-                                virtual void append(FFoption option, const IFieldFilter * filter) override
-                                {
-                                    fieldFilters.append(*filter);
-                                }
-                            } context(fieldFilters);
-                            helper->createSegmentMonitors(&context);
-
-                            RowFilter actualFilter;
-                            Owned<const IKeyTranslator> keyedTranslator = createKeyTranslator(actualFormat->queryRecordAccessor(true), expectedFormat->queryRecordAccessor(true));
-                            if (keyedTranslator && keyedTranslator->needsTranslate())
-                                keyedTranslator->translate(actualFilter, fieldFilters);
-                            else
-                                actualFilter.appendFilters(fieldFilters);
-
-                            Owned<IIndexLookup> indexLookup = createRemoteFilteredKey(ep, lPath, crc, actualFormat, projectedFormat, actualFilter, remoteLimit);
-                            if (indexLookup)
-                            {
-                                try
-                                {
-                                    indexLookup->ensureAvailable();
-                                }
-                                catch (IException *e)
-                                {
-                                    EXCLOG(e, nullptr);
-                                    e->Release();
-                                    continue; // try next copy and ultimately failover to local when no more copies
-                                }
-                                ActPrintLog("[part=%d]: reading remote dafilesrv index '%s' (logical file = %s)", partNum, path.str(), logicalFilename.get());
-                                partNum = p;
-                                return indexLookup.getClear();
                             }
+                            virtual void append(IKeySegmentMonitor *segment) override { throwUnexpected(); }
+                            virtual void append(FFoption option, const IFieldFilter * filter) override
+                            {
+                                fieldFilters.append(*filter);
+                            }
+                        } context(fieldFilters);
+                        helper->createSegmentMonitors(&context);
+
+                        RowFilter actualFilter;
+                        Owned<const IKeyTranslator> keyedTranslator = createKeyTranslator(actualFormat->queryRecordAccessor(true), expectedFormat->queryRecordAccessor(true));
+                        if (keyedTranslator && keyedTranslator->needsTranslate())
+                            keyedTranslator->translate(actualFilter, fieldFilters);
+                        else
+                            actualFilter.appendFilters(fieldFilters);
+
+                        StringBuffer lPath;
+                        rfn.getLocalPath(lPath);
+                        Owned<IIndexLookup> indexLookup = createRemoteFilteredKey(ep, lPath, crc, actualFormat, projectedFormat, actualFilter, remoteLimit);
+                        if (indexLookup)
+                        {
+                            try
+                            {
+                                indexLookup->ensureAvailable();
+                            }
+                            catch (IException *e)
+                            {
+                                EXCLOG(e, nullptr);
+                                e->Release();
+                                continue; // try next copy and ultimately failover to local when no more copies
+                            }
+                            ActPrintLog("[part=%d]: reading remote dafilesrv index '%s' (logical file = %s)", partNum, path.str(), logicalFilename.get());
+                            partNum = p;
+                            return indexLookup.getClear();
                         }
                     }
                 }

